presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewPlanValidator.java

```diff
@@ -34,15 +42,36 @@
 {
     private final Node query;
 
+    private final boolean analyzeColumnMapping;
+    private final Optional<Analysis> analysis;
+
+    private ImmutableList.Builder<List<TableColumn>> linkedBaseColumns;
+
     public MaterializedViewPlanValidator(Node query)
     {
+        this(query, false, Optional.empty());
+    }
+
+    public MaterializedViewPlanValidator(Node query, boolean analyzeColumnMapping, Optional<Analysis> analysis)
+    {
+        checkArgument(!analyzeColumnMapping || analysis.isPresent(), "analysis must be set if column mapping is to be computed.");
+
         this.query = requireNonNull(query, "query is null");
+        this.analyzeColumnMapping = analyzeColumnMapping;
+        this.analysis = requireNonNull(analysis, "analysis is null");
+
+        this.linkedBaseColumns = ImmutableList.builder();
+    }
+
+    public List<List<TableColumn>> getLinkedBaseColumns()
+    {
+        return linkedBaseColumns.build();
     }
```

Comment:
This is again an anti-pattern:
- the validator should not be stateful; it should just tell you true or false
- what if we run the validator twice? This builder will keep growing.

Reply:
Good point. I can revise. (Just there're some examples doing it this way in the codebase)

Reply:
Revised by making the visitor private and wrapping it with public interfaces. So anyone could only access the class through the public interfaces and we will control how the visitor is called. Also moved the the linkedBaseColumns to the Context instead of the Visitor.